firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +60,142 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  static class CallbackList {
+    void run() {
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void add(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private final List<Runnable> callbacks = new ArrayList<>();
+  }
+
+  public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    CallbackList callbacks = new CallbackList();
+  }
+
+  @SuppressWarnings("deprecation")
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    CallbackList callbacks = new CallbackList();
+  }
+
+  @Nullable
+  private static <T> T castFragment(Class<T> fragmentClass, @Nullable Object fragment, String tag) {
+    try {
+      return fragmentClass.cast(fragment);
+    } catch (ClassCastException e) {
+      throw new IllegalStateException(
+          "Fragment with tag '"
+              + tag
+              + "' is a "
+              + fragment.getClass().getName()
+              + " but should be a "
+              + fragmentClass.getName());
+    }
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private static WeakHashMap<Activity, WeakReference<StopListenerSupportFragment>>
+      supportFragmentByActivity = new WeakHashMap<>();
+  private static WeakHashMap<Activity, WeakReference<StopListenerFragment>> fragmentByActivity =
+      new WeakHashMap<>();
+
+  /**
+   * Implementation for non-FragmentActivity Activity's. Unfortunatly, all Fragment related
+   * classes/methods with nonFragmentActivityActivity's are deprecated, imply that almost everything
+   * in this function is deprecated.
+   */
+  @SuppressWarnings("deprecation")
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    hardAssert(
+        !(activity instanceof FragmentActivity),
+        "onActivityStopCallOnce must be called with a *non*-FragmentActivity Activity.");
+
+    synchronized (activity) {
+      StopListenerFragment fragment;
+
+      WeakReference<StopListenerFragment> ref = fragmentByActivity.get(activity);
```

Comment:
This is potentially problematic because activity is of a user-controlled type which could override equals/hashCode in broken ways. Most problematically, the properties of the key object that contribute to its equality/hash code must be immutable for the lifetime of the object or you horribly break java Maps.

If equals/hashCode are final in Activity this may not be an issue but if they're user controlled this isn't a good hash key.

Guava's MapMaker could be a reasonable answer: MapMaker().weakKeys().makeMap() does the right thing, specifically using object identity rather than equals/hashCode 

https://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/MapMaker.html